
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://monarch-initiative.github.io/ontogpt/">
      
      
      
        <link rel="next" href="core/">
      
      <link rel="icon" href="assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.2, mkdocs-material-9.0.10">
    
    
      
        <title>OntoGPT</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.0d440cfe.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.2505c338.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="amber">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#ontogpt" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="." title="OntoGPT" class="md-header__button md-logo" aria-label="OntoGPT" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            OntoGPT
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              About
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="amber"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 0-7 7c0 2.38 1.19 4.47 3 5.74V17a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1v-2.26c1.81-1.27 3-3.36 3-5.74a7 7 0 0 0-7-7M9 21a1 1 0 0 0 1 1h4a1 1 0 0 0 1-1v-1H9v1Z"/></svg>
            </label>
          
        
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="amber"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 1 7 7c0 2.38-1.19 4.47-3 5.74V17a1 1 0 0 1-1 1H9a1 1 0 0 1-1-1v-2.26C6.19 13.47 5 11.38 5 9a7 7 0 0 1 7-7M9 21v-1h6v1a1 1 0 0 1-1 1h-4a1 1 0 0 1-1-1m3-17a5 5 0 0 0-5 5c0 2.05 1.23 3.81 3 4.58V16h4v-2.42c1.77-.77 3-2.53 3-4.58a5 5 0 0 0-5-5Z"/></svg>
            </label>
          
        
      </form>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/monarch-initiative/ontogpt/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="." title="OntoGPT" class="md-nav__button md-logo" aria-label="OntoGPT" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    OntoGPT
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/monarch-initiative/ontogpt/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          About
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="." class="md-nav__link md-nav__link--active">
        About
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#spires-usage" class="md-nav__link">
    SPIRES: Usage
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-it-works" class="md-nav__link">
    How it works
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pre-requisites" class="md-nav__link">
    Pre-requisites
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-define-your-own-extraction-data-model" class="md-nav__link">
    How to define your own extraction data model
  </a>
  
    <nav class="md-nav" aria-label="How to define your own extraction data model">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-1-define-a-schema" class="md-nav__link">
    Step 1: Define a schema
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-2-compile-the-schema" class="md-nav__link">
    Step 2: Compile the schema
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-3-run-the-command-line" class="md-nav__link">
    Step 3: Run the command line
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#web-application" class="md-nav__link">
    Web Application
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#features" class="md-nav__link">
    Features
  </a>
  
    <nav class="md-nav" aria-label="Features">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#multiple-levels-of-nesting" class="md-nav__link">
    Multiple Levels of nesting
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#text-length-limit" class="md-nav__link">
    Text length limit
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#haloe-usage" class="md-nav__link">
    HALOE: Usage
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#limitations" class="md-nav__link">
    Limitations
  </a>
  
    <nav class="md-nav" aria-label="Limitations">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#non-deterministic" class="md-nav__link">
    Non-deterministic
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#coupled-to-openai" class="md-nav__link">
    Coupled to OpenAI
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
      
      
      
        <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
          SPIRES Templates
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          SPIRES Templates
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="core/" class="md-nav__link">
        Core Upper Level Schema
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="reaction/" class="md-nav__link">
        Biochemical Reactions
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="biological_process/" class="md-nav__link">
        Biological Processes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="cell_type/" class="md-nav__link">
        Cell Types
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="ctd/" class="md-nav__link">
        Chemical to Disease
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="diagnostic_procedure/" class="md-nav__link">
        Diagnostic Procedures
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="drug/" class="md-nav__link">
        Drugs and Mechanisms
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="environmental_sample/" class="md-nav__link">
        Environmental Samples
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="mendelian_disease/" class="md-nav__link">
        Mendelian Diseases
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="metabolic_process/" class="md-nav__link">
        Metabolic Processes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="metagenome_study/" class="md-nav__link">
        Metagenome Studies
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="gocam/" class="md-nav__link">
        Pathways
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="recipe/" class="md-nav__link">
        Recipes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="treatment/" class="md-nav__link">
        Treatments
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="halo/" class="md-nav__link">
        HALO Schema
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#spires-usage" class="md-nav__link">
    SPIRES: Usage
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-it-works" class="md-nav__link">
    How it works
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pre-requisites" class="md-nav__link">
    Pre-requisites
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-define-your-own-extraction-data-model" class="md-nav__link">
    How to define your own extraction data model
  </a>
  
    <nav class="md-nav" aria-label="How to define your own extraction data model">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-1-define-a-schema" class="md-nav__link">
    Step 1: Define a schema
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-2-compile-the-schema" class="md-nav__link">
    Step 2: Compile the schema
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-3-run-the-command-line" class="md-nav__link">
    Step 3: Run the command line
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#web-application" class="md-nav__link">
    Web Application
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#features" class="md-nav__link">
    Features
  </a>
  
    <nav class="md-nav" aria-label="Features">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#multiple-levels-of-nesting" class="md-nav__link">
    Multiple Levels of nesting
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#text-length-limit" class="md-nav__link">
    Text length limit
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#haloe-usage" class="md-nav__link">
    HALOE: Usage
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#limitations" class="md-nav__link">
    Limitations
  </a>
  
    <nav class="md-nav" aria-label="Limitations">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#non-deterministic" class="md-nav__link">
    Non-deterministic
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#coupled-to-openai" class="md-nav__link">
    Coupled to OpenAI
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  



<h1 id="ontogpt">OntoGPT</h1>
<p>Generation of Ontologies and Knowledge Bases using GPT</p>
<p>A knowledge extraction tool that uses a large language model (LLM) to extract semantic information from text.</p>
<p>This exploits the ability of LLMs such as GPT-3 to return user-defined data structures as a response.</p>
<p>Currently there are two different pipelines implemented:</p>
<ul>
<li>SPIRES: Structured Prompt Interrogation and Recursive Extraction of Semantics<ul>
<li>Zero-shot learning approach to extracting nested semantic structures from text</li>
<li>Inputs: LinkML schema + text</li>
<li>Outputs: JSON, YAML, or RDF or OWL that conforms to the schema</li>
</ul>
</li>
<li>HALO: HAllucinating Latent Ontologies <ul>
<li>Few-shot learning approach to generating/hallucinating a domain ontology given a few examples</li>
</ul>
</li>
</ul>
<h2 id="spires-usage">SPIRES: Usage</h2>
<p>Given a short text <code>abstract.txt</code> with content such as:</p>
<blockquote>
<p>The cGAS/STING-mediated DNA-sensing signaling pathway is crucial
   for interferon (IFN) production and host antiviral
   responses</p>
<p>...
[snip] 
...</p>
<p>The underlying mechanism was the
   interaction of US3 with β-catenin and its hyperphosphorylation of
   β-catenin at Thr556 to block its nuclear translocation
...
...</p>
</blockquote>
<p>(see <a href="https://github.com/monarch-initiative/ontogpt/blob/main/tests/input/cases/gocam-betacat.txt">full input</a>)</p>
<p>We can extract this into the <a href="https://github.com/monarch-initiative/ontogpt/blob/main/src/ontogpt/templates/gocam.yaml">GO pathway datamodel</a>:</p>
<pre><code class="language-bash">ontogpt extract -t gocam.GoCamAnnotations abstract.txt
</code></pre>
<p>Giving schema-compliant YAML such as:</p>
<pre><code class="language-yaml">genes:
- HGNC:2514
- HGNC:21367
- HGNC:27962
- AUTO:US3
- FPLX:Interferon
- AUTO:ISG
gene_gene_interactions:
- gene1: AUTO:US3
  gene2: HGNC:2514
gene_localizations:
- gene: HGNC:2514
  location: Nuclear
gene_functions:
- gene: HGNC:2514
  molecular_activity: Transcription
- gene: HGNC:21367
  molecular_activity: Production
...
</code></pre>
<p>See <a href="https://github.com/monarch-initiative/ontogpt/blob/main/tests/output/gocam-betacat.yaml">full output</a>.</p>
<p>Ungrounded named entities appear as text preceded by AUTO (or your preferred prefix, provided with the --auto-prefix option when using the extract command).</p>
<h2 id="how-it-works">How it works</h2>
<ol>
<li>You provide an arbitrary data model, describing the structure you want to extract text into<ul>
<li>this can be nested (but see limitations below)</li>
</ul>
</li>
<li>Provide your preferred annotations for grounding NamedEntity fields</li>
<li>OntoGPT will:<ul>
<li>generate a prompt</li>
<li>feed the prompt to a language model (currently one of OpenAI's models)</li>
<li>parse the results into a dictionary structure</li>
<li>ground the results using a preferred annotator</li>
</ul>
</li>
</ol>
<h2 id="pre-requisites">Pre-requisites</h2>
<ul>
<li>Python 3.9+</li>
<li>an OpenAI account</li>
<li>a BioPortal account (optional, for grounding)</li>
</ul>
<p>You will need to set both API keys using OAK (which is a dependency of this project) as:</p>
<pre><code>poetry run runoak set-apikey -e openai &lt;your openai api key&gt;
poetry run runoak set-apikey -e bioportal &lt;your bioportal api key&gt;
</code></pre>
<h2 id="how-to-define-your-own-extraction-data-model">How to define your own extraction data model</h2>
<h3 id="step-1-define-a-schema">Step 1: Define a schema</h3>
<p>See <a href="https://github.com/monarch-initiative/ontogpt/tree/main/src/ontogpt/templates">src/ontogpt/templates/</a> for examples.</p>
<p>Define a schema (using a subset of <a href="https://linkml.io/">LinkML</a>) that describes the structure you want to extract from your text.</p>
<pre><code class="language-yaml">classes:
  MendelianDisease:
    attributes:
      name:
        description: the name of the disease
        examples:
          - value: peroxisome biogenesis disorder
        identifier: true  ## needed for inlining
      description:
        description: a description of the disease
        examples:
          - value: &gt;-
             Peroxisome biogenesis disorders, Zellweger syndrome spectrum (PBD-ZSS) is a group of autosomal recessive disorders affecting the formation of functional peroxisomes, characterized by sensorineural hearing loss, pigmentary retinal degeneration, multiple organ dysfunction and psychomotor impairment
      synonyms:
        multivalued: true
        examples:
          - value: Zellweger syndrome spectrum
          - value: PBD-ZSS
      subclass_of:
        multivalued: true
        range: MendelianDisease
        examples:
          - value: lysosomal disease
          - value: autosomal recessive disorder
      symptoms:
        range: Symptom
        multivalued: true
        examples:
          - value: sensorineural hearing loss
          - value: pigmentary retinal degeneration
      inheritance:
        range: Inheritance
        examples:
          - value: autosomal recessive
      genes:
        range: Gene
        multivalued: true
        examples:
          - value: PEX1
          - value: PEX2
          - value: PEX3

  Gene:
    is_a: NamedThing
    id_prefixes:
      - HGNC
    annotations:
      annotators: gilda:, bioportal:hgnc-nr

  Symptom:
    is_a: NamedThing
    id_prefixes:
      - HP
    annotations:
      annotators: sqlite:obo:hp

  Inheritance:
    is_a: NamedThing
    annotations:
      annotators: sqlite:obo:hp
</code></pre>
<ul>
<li>The schema is defined in LinkML.</li>
<li>Prompt hints can be specified using the <code>prompt</code> annotation (otherwise description is used).</li>
<li>Multivalued fields are supported.</li>
<li>The default range is string - these are not grounded, e.g., disease name, synonyms.</li>
<li>Define a class for each NamedEntity.</li>
<li>For any NamedEntity, you can specify one or more preferred annotators using the <code>annotators</code> annotation.</li>
</ul>
<p>We recommend following an established schema like <a href="https://biolink.github.io/biolink-model/">Biolink Model</a>, but you can define your own.</p>
<h3 id="step-2-compile-the-schema">Step 2: Compile the schema</h3>
<p>Run the <code>make</code> command at the top level. This will compile the schemas in the templates directory to Pydantic classes.</p>
<p>Alternatively, run <code>make src/templates/{name_of_schema}.py</code> to compile just this schema.
For example, if your schema is defined within <code>entities.yaml</code>, run <code>make src/templates/entities.py</code>.</p>
<h3 id="step-3-run-the-command-line">Step 3: Run the command line</h3>
<p>e.g.</p>
<pre><code>ontogpt extract -t mendelian_disease.MendelianDisease marfan-wikipedia.txt
</code></pre>
<h2 id="web-application">Web Application</h2>
<p>There is a bare bones web application for running OntoGPT and viewing results.</p>
<p>Install the required dependencies first with the following command:</p>
<pre><code>poetry install -E web
</code></pre>
<p>Then run the following command to start the web application:</p>
<pre><code>poetry run web-ontogpt
</code></pre>
<p>Note that the agent running uvicorn must have the API key set, so for obvious reasons
don't host this publicly without authentication, unless you want your credits drained.</p>
<h2 id="features">Features</h2>
<h3 id="multiple-levels-of-nesting">Multiple Levels of nesting</h3>
<p>Currently no more than two levels of nesting are recommended.</p>
<p>If a field has a range which is itself a class and not a primitive, it will attempt to nest</p>
<p>E.g. the <a href="http://geneontology.org/docs/gocam-overview/">GO-CAM</a> schema has an attribute:</p>
<pre><code class="language-yaml">  attributes:
      ...
      gene_functions:
        description: semicolon-separated list of gene to molecular activity relationships
        multivalued: true
        range: GeneMolecularActivityRelationship
</code></pre>
<p>Because GeneMolecularActivityRelationship is <em>inlined</em> it will nest.</p>
<p>The generated prompt is:</p>
<p><code>gene_functions : &lt;semicolon-separated list of gene to molecular activities relationships&gt;</code></p>
<p>The output of this is then passed through further llama iterations.</p>
<h2 id="text-length-limit">Text length limit</h2>
<p>Currently SPIRES must use text-davinci-003, which has a total 4k token limit (prompt + completion).</p>
<p>You can pass in a parameter to split the text into chunks, results will be recombined automatically,
but more experiments need to be done to determined how reliable this is.</p>
<p>```</p>
<h2 id="haloe-usage">HALOE: Usage</h2>
<p>TODO</p>
<h2 id="limitations">Limitations</h2>
<h3 id="non-deterministic">Non-deterministic</h3>
<p>This relies on an existing LLM, and LLMs can be fickle in their responses.</p>
<h3 id="coupled-to-openai">Coupled to OpenAI</h3>
<p>At present, you will need an OpenAI account.</p>
<h1 id="acknowledgements">Acknowledgements</h1>
<p>This <a href="https://cookiecutter.readthedocs.io/en/stable/README.html">cookiecutter</a> project was developed from the <a href="https://github.com/hrshdhgd/sphintoxetry-cookiecutter">sphintoxetry-cookiecutter</a> template and will be kept up-to-date using <a href="https://cruft.github.io/cruft/">cruft</a>.</p>





                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": ".", "features": ["content.tabs.link"], "search": "assets/javascripts/workers/search.db81ec45.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="assets/javascripts/bundle.6df46069.min.js"></script>
      
    
  </body>
</html>