
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://monarch-initiative.github.io/ontogpt/">
      
      
      
        <link rel="next" href="core/">
      
      <link rel="icon" href="assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.2, mkdocs-material-9.0.10">
    
    
      
        <title>OntoGPT</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.0d440cfe.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.2505c338.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="amber">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#ontogpt" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="." title="OntoGPT" class="md-header__button md-logo" aria-label="OntoGPT" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            OntoGPT
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              About
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="amber"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 0-7 7c0 2.38 1.19 4.47 3 5.74V17a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1v-2.26c1.81-1.27 3-3.36 3-5.74a7 7 0 0 0-7-7M9 21a1 1 0 0 0 1 1h4a1 1 0 0 0 1-1v-1H9v1Z"/></svg>
            </label>
          
        
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="amber"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2a7 7 0 0 1 7 7c0 2.38-1.19 4.47-3 5.74V17a1 1 0 0 1-1 1H9a1 1 0 0 1-1-1v-2.26C6.19 13.47 5 11.38 5 9a7 7 0 0 1 7-7M9 21v-1h6v1a1 1 0 0 1-1 1h-4a1 1 0 0 1-1-1m3-17a5 5 0 0 0-5 5c0 2.05 1.23 3.81 3 4.58V16h4v-2.42c1.77-.77 3-2.53 3-4.58a5 5 0 0 0-5-5Z"/></svg>
            </label>
          
        
      </form>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/monarch-initiative/ontogpt/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="." title="OntoGPT" class="md-nav__button md-logo" aria-label="OntoGPT" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    OntoGPT
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/monarch-initiative/ontogpt/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          About
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="." class="md-nav__link md-nav__link--active">
        About
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    Introduction
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pre-requisites" class="md-nav__link">
    Pre-requisites
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    Setup
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#examples" class="md-nav__link">
    Examples
  </a>
  
    <nav class="md-nav" aria-label="Examples">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#strategy-1-knowledge-extraction-using-spires" class="md-nav__link">
    Strategy 1: Knowledge extraction using SPIRES
  </a>
  
    <nav class="md-nav" aria-label="Strategy 1: Knowledge extraction using SPIRES">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input" class="md-nav__link">
    Input
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#command" class="md-nav__link">
    Command
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    Output
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#working-mechanism" class="md-nav__link">
    Working Mechanism
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gene-enrichment-using-spindoctor" class="md-nav__link">
    Gene Enrichment using SPINDOCTOR
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#features" class="md-nav__link">
    Features
  </a>
  
    <nav class="md-nav" aria-label="Features">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#define-your-own-extraction-model-using-linkml" class="md-nav__link">
    Define your own extraction model using LinkML
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#multiple-levels-of-nesting" class="md-nav__link">
    Multiple levels of nesting
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#text-length-limit" class="md-nav__link">
    Text length limit
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#schema-tips" class="md-nav__link">
    Schema tips
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#owl-exports" class="md-nav__link">
    OWL Exports
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#web-application-setup" class="md-nav__link">
    Web Application Setup
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ontogpt-limitations" class="md-nav__link">
    OntoGPT Limitations
  </a>
  
    <nav class="md-nav" aria-label="OntoGPT Limitations">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#spindoctor-web-app" class="md-nav__link">
    SPINDOCTOR web app
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#huggingface-hub" class="md-nav__link">
    HuggingFace Hub
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-local-models" class="md-nav__link">
    Using local models
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#citation" class="md-nav__link">
    Citation
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#contributing" class="md-nav__link">
    Contributing
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#acknowledgements" class="md-nav__link">
    Acknowledgements
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
      
      
      
        <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
          SPIRES Templates
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          SPIRES Templates
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="core/" class="md-nav__link">
        Core Upper Level Schema
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="reaction/" class="md-nav__link">
        Biochemical Reactions
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="biological_process/" class="md-nav__link">
        Biological Processes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="cell_type/" class="md-nav__link">
        Cell Types
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="ctd/" class="md-nav__link">
        Chemical to Disease
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="diagnostic_procedure/" class="md-nav__link">
        Diagnostic Procedures
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="drug/" class="md-nav__link">
        Drugs and Mechanisms
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="environmental_sample/" class="md-nav__link">
        Environmental Samples
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="mendelian_disease/" class="md-nav__link">
        Mendelian Diseases
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="metabolic_process/" class="md-nav__link">
        Metabolic Processes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="metagenome_study/" class="md-nav__link">
        Metagenome Studies
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="gocam/" class="md-nav__link">
        Pathways
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="recipe/" class="md-nav__link">
        Recipes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="treatment/" class="md-nav__link">
        Treatments
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="halo/" class="md-nav__link">
        HALO Schema
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    Introduction
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pre-requisites" class="md-nav__link">
    Pre-requisites
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    Setup
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#examples" class="md-nav__link">
    Examples
  </a>
  
    <nav class="md-nav" aria-label="Examples">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#strategy-1-knowledge-extraction-using-spires" class="md-nav__link">
    Strategy 1: Knowledge extraction using SPIRES
  </a>
  
    <nav class="md-nav" aria-label="Strategy 1: Knowledge extraction using SPIRES">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input" class="md-nav__link">
    Input
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#command" class="md-nav__link">
    Command
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    Output
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#working-mechanism" class="md-nav__link">
    Working Mechanism
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gene-enrichment-using-spindoctor" class="md-nav__link">
    Gene Enrichment using SPINDOCTOR
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#features" class="md-nav__link">
    Features
  </a>
  
    <nav class="md-nav" aria-label="Features">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#define-your-own-extraction-model-using-linkml" class="md-nav__link">
    Define your own extraction model using LinkML
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#multiple-levels-of-nesting" class="md-nav__link">
    Multiple levels of nesting
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#text-length-limit" class="md-nav__link">
    Text length limit
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#schema-tips" class="md-nav__link">
    Schema tips
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#owl-exports" class="md-nav__link">
    OWL Exports
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#web-application-setup" class="md-nav__link">
    Web Application Setup
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ontogpt-limitations" class="md-nav__link">
    OntoGPT Limitations
  </a>
  
    <nav class="md-nav" aria-label="OntoGPT Limitations">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#spindoctor-web-app" class="md-nav__link">
    SPINDOCTOR web app
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#huggingface-hub" class="md-nav__link">
    HuggingFace Hub
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-local-models" class="md-nav__link">
    Using local models
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#citation" class="md-nav__link">
    Citation
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#contributing" class="md-nav__link">
    Contributing
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#acknowledgements" class="md-nav__link">
    Acknowledgements
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  



<h1 id="ontogpt">OntoGPT</h1>
<h2 id="introduction">Introduction</h2>
<p>OntoGPT is a Python package for the generation of Ontologies and Knowledge Bases using GPT. It is a knowledge extraction tool that uses a Large Language Models (LLMs) to extract semantic information from text.</p>
<p>This makes use of so-called <em>instruction prompts</em> in Large Language Models (LLMs) such as GPT-4.</p>
<p>Currently three different strategies for knowledge extraction have been implemented in the ontogpt package:</p>
<ul>
<li>SPIRES: <em>Structured Prompt Interrogation and Recursive Extraction of Semantics</em></li>
<li>Zero-shot learning (ZSL) approach to extracting nested semantic structures from text</li>
<li>This approach takes two inputs - 1) LinkML schema 2) free text, and outputs knowledge in a structure conformant with the supplied schema in JSON, YAML, RDF or OWL formats</li>
<li>Uses text-davinci-003 or gpt-3.5-turbo (gpt-4 untested)</li>
<li>HALO: <em>HAllucinating Latent Ontologies</em></li>
<li>Few-shot learning approach to generating/hallucinating a domain ontology given a few examples</li>
<li>Uses code-davinci-002</li>
<li>SPINDOCTOR: <em>Structured Prompt Interpolation of Narrative Descriptions Or Controlled Terms for Ontological Reporting</em></li>
<li>Summarize gene set descriptions (pseudo gene-set enrichment)</li>
<li>Uses text-davinci-003 or gpt-3.5-turbo (gpt-4 untested)</li>
</ul>
<h2 id="pre-requisites">Pre-requisites</h2>
<ul>
<li>Python 3.9+</li>
<li>OpenAI account</li>
</ul>
<p>An OpenAI key is necessary for using OpenAI's GPT models. This is a paid API and you will be charged based on usage. If you do not have an OpenAI account, <a href="https://platform.openai.com/signup">you may sign up here</a>. You will need to set your API key using the <a href="https://github.com/INCATools/ontology-access-kit">Ontology Access Kit</a>:</p>
<pre><code class="language-bash">poetry run runoak set-apikey -e openai &lt;your openai api key&gt;
</code></pre>
<p>You may also set additional API keys for optional resources:</p>
<ul>
<li><a href="https://bioportal.bioontology.org/">BioPortal</a> account (for grounding). The BioPortal key is necessary for using ontologies from <a href="https://bioportal.bioontology.org/">BioPortal</a>. You may get a key by signing up for an account on their web site.</li>
<li><a href="https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/">NCBI E-utilities</a>. The NCBI email address and API key are used for retrieving text and metadata from PubMed. You may still access these resources without identifying yourself, but you may encounter rate limiting and errors.</li>
<li><a href="https://huggingface.co/docs/api-inference/quicktour#get-your-api-token">HuggingFace Hub</a>. This API key is necessary to retrieve models from the HuggingFace Hub service.</li>
</ul>
<p>These optional keys may be set as follows:</p>
<pre><code class="language-bash">poetry run runoak set-apikey -e bioportal &lt;your bioportal api key&gt;
poetry run runoak set-apikey -e ncbi-email &lt;your email address&gt;
poetry run runoak set-apikey -e ncbi-key &lt;your NCBI api key&gt;
poetry run runoak set-apikey -e hfhub-key &lt;your HuggingFace Hub api key&gt;
</code></pre>
<h2 id="setup">Setup</h2>
<p>For feature development and contributing to the package:</p>
<pre><code class="language-bash">git clone https://github.com/monarch-initiative/ontogpt.git
cd ~/path/to/ontogpt
poetry install
</code></pre>
<p>To simply start using the package in your workspace:</p>
<pre><code class="language-bash">pip install ontogpt
</code></pre>
<p>Note that some features require installing additional, optional dependencies.</p>
<p>These may be installed as:</p>
<pre><code class="language-bash">poetry install --extras extra_name
# OR
pip install ontogpt[extra_name]
</code></pre>
<p>where <code>extra_name</code> is one of the following:</p>
<ul>
<li><code>docs</code> - dependencies for building documentation</li>
<li><code>web</code> - dependencies for the web application</li>
<li><code>recipes</code> - dependencies for recipe scraping and parsing</li>
<li><code>gpt4all</code> - dependencies for loading LLMs from GPT4All</li>
<li><code>textract</code> - the textract plugin</li>
<li><code>huggingface</code> - dependencies for accessing LLMs from HuggingFace Hub, remotely or locally</li>
</ul>
<h2 id="examples">Examples</h2>
<h3 id="strategy-1-knowledge-extraction-using-spires">Strategy 1: Knowledge extraction using SPIRES</h3>
<h4 id="input">Input</h4>
<p>Consider some text from one of the input files being used in the ontogpt test suite. You can find the text file <a href="tests/input/cases/gocam-betacat.txt">here</a>. You can download the raw file from the GitHub link to that input text file, or copy its contents over into another file, say, <code>abstract.txt</code>. An excerpt </p>
<blockquote>
<p>The cGAS/STING-mediated DNA-sensing signaling pathway is crucial
  for interferon (IFN) production and host antiviral
  responses</p>
<p>...
[snip]
...</p>
<p>The underlying mechanism was the
  interaction of US3 with β-catenin and its hyperphosphorylation of
  β-catenin at Thr556 to block its nuclear translocation
...
...</p>
</blockquote>
<p>We can extract knowledge from the above text this into the <a href="src/ontogpt/templates/gocam.yaml">GO pathway datamodel</a> by running the following command:</p>
<h4 id="command">Command</h4>
<pre><code class="language-bash">ontogpt extract -t gocam.GoCamAnnotations -i ~/path/to/abstract.txt
</code></pre>
<p>Note: The value accepted by the <code>-t</code> / <code>--template</code> argument is the base name of one of the LinkML schema / data model which can be found in the <a href="src/ontogpt/templates/">templates</a> folder.</p>
<h4 id="output">Output</h4>
<p>The output returned from the above command can be optionally redirected into an output file using the <code>-o</code> / <code>--output</code>.</p>
<p>The following is a small part of what the larger schema-compliant output looks like:</p>
<pre><code class="language-yaml">genes:
- HGNC:2514
- HGNC:21367
- HGNC:27962
- US3
- FPLX:Interferon
- ISG
gene_gene_interactions:
- gene1: US3
  gene2: HGNC:2514
gene_localizations:
- gene: HGNC:2514
  location: Nuclear
gene_functions:
- gene: HGNC:2514
  molecular_activity: Transcription
- gene: HGNC:21367
  molecular_activity: Production
...
</code></pre>
<h4 id="working-mechanism">Working Mechanism</h4>
<ol>
<li>You provide an arbitrary data model, describing the structure you want to extract text into. This can be nested (but see limitations below)</li>
<li>Provide your preferred annotations for grounding <code>NamedEntity</code> fields</li>
<li>
<p>OntoGPT will:</p>
</li>
<li>
<p>Generate a prompt</p>
</li>
<li>Feed the prompt to a language model (currently OpenAI GPT models)</li>
<li>Parse the results into a dictionary structure</li>
<li>Ground the results using a preferred annotator</li>
</ol>
<h2 id="gene-enrichment-using-spindoctor">Gene Enrichment using SPINDOCTOR</h2>
<p>Given a set of genes, OntoGPT can find similarities among them.</p>
<p>Ex.:</p>
<pre><code class="language-bash">ontogpt enrichment -U tests/input/genesets/sensory-ataxia.yaml
</code></pre>
<p>The default is to use ontological gene function synopses (via the Alliance API).</p>
<ul>
<li>To use narrative/RefSeq summaries, use the <code>--no-ontological-synopses</code> flag</li>
<li>To run without any gene descriptions, use the <code>--no-annotations</code> flag</li>
</ul>
<h2 id="features">Features</h2>
<h3 id="define-your-own-extraction-model-using-linkml">Define your own extraction model using LinkML</h3>
<p>There are a number of pre-defined LinkML data models already developed here - <a href="src/ontogpt/templates/">src/ontogpt/templates/</a> which you can use as reference when creating your own data models.</p>
<p>Define a schema (using a subset of <a href="https://linkml.io">LinkML</a>) that describes the structure in which you want to extract knowledge from your text.</p>
<p>```yaml
  classes:
    MendelianDisease:
      attributes:
        name:
          description: the name of the disease
          examples:
            - value: peroxisome biogenesis disorder
          identifier: true  ## needed for inlining
        description:
          description: a description of the disease
          examples:
            - value: &gt;-
              Peroxisome biogenesis disorders, Zellweger syndrome spectrum (PBD-ZSS) is a group of autosomal recessive disorders affecting the formation of functional peroxisomes, characterized by sensorineural hearing loss, pigmentary retinal degeneration, multiple organ dysfunction and psychomotor impairment
        synonyms:
          multivalued: true
          examples:
            - value: Zellweger syndrome spectrum
            - value: PBD-ZSS
        subclass_of:
          multivalued: true
          range: MendelianDisease
          examples:
            - value: lysosomal disease
            - value: autosomal recessive disorder
        symptoms:
          range: Symptom
          multivalued: true
          examples:
            - value: sensorineural hearing loss
            - value: pigmentary retinal degeneration
        inheritance:
          range: Inheritance
          examples:
            - value: autosomal recessive
        genes:
          range: Gene
          multivalued: true
          examples:
            - value: PEX1
            - value: PEX2
            - value: PEX3</p>
<pre><code>Gene:
  is_a: NamedThing
  id_prefixes:
    - HGNC
  annotations:
    annotators: gilda:, bioportal:hgnc-nr

Symptom:
  is_a: NamedThing
  id_prefixes:
    - HP
  annotations:
    annotators: sqlite:obo:hp

Inheritance:
  is_a: NamedThing
  annotations:
    annotators: sqlite:obo:hp
```
</code></pre>
<ul>
<li>Prompt hints can be specified using the <code>prompt</code> annotation (otherwise description is used)</li>
<li>Multivalued fields are supported</li>
<li>The default range is string — these are not grounded. Ex.: disease name, synonyms</li>
<li>Define a class for each <code>NamedEntity</code></li>
<li>For any <code>NamedEntity</code>, you can specify a preferred annotator using the <code>annotators</code> annotation</li>
</ul>
<p>We recommend following an established schema like <a href="https://github.com/biolink/biolink-model">BioLink Model</a>, but you can define your own.</p>
<p>Next step is to compile the schema. For that, you should place the schema YAML in the directory <a href="src/ontogpt/templates/">src/ontogpt/templates/</a>. Then, run the <code>make</code> command at the top level. This will compile the schema to Python (Pydantic classes).</p>
<p>Once you have defined your own schema / data model and placed in the correct directory, you can run the <code>extract</code> command. </p>
<p>Ex.:</p>
<pre><code>ontogpt extract -t mendelian_disease.MendelianDisease -i marfan-wikipedia.txt
</code></pre>
<h3 id="multiple-levels-of-nesting">Multiple levels of nesting</h3>
<p>Currently no more than two levels of nesting are recommended.</p>
<p>If a field has a range which is itself a class and not a primitive, it will attempt to nest.</p>
<p>Ex. the <code>gocam</code> schema has an attribute:</p>
<pre><code class="language-yaml">  attributes:
      ...
      gene_functions:
        description: semicolon-separated list of gene to molecular activity relationships
        multivalued: true
        range: GeneMolecularActivityRelationship
</code></pre>
<p>The range <code>GeneMolecularActivityRelationship</code> has been specified <em>inline</em>, so it will nest.</p>
<p>The generated prompt is:</p>
<pre><code class="language-bash">gene_functions : &lt;semicolon-separated list of gene to molecular activities relationships&gt;
</code></pre>
<p>The output of this is then passed through further SPIRES iterations.</p>
<h3 id="text-length-limit">Text length limit</h3>
<p>Currently SPIRES must use text-davinci-003, which has a total 4k token limit (prompt + completion).</p>
<p>You can pass in a parameter to split the text into chunks. Returned results will be recombined automatically, but more experiments need to be done to determined how reliable this is.</p>
<h3 id="schema-tips">Schema tips</h3>
<p>It helps to have an understanding of the <a href="https://linkml.io">LinkML</a> schema language, but it should be possible to define your own schemas using the examples in <a href="src/ontogpt/templates/">src/ontogpt/templates</a> as a guide.</p>
<p>OntoGPT-specific extensions are specified as <em>annotations</em>.</p>
<p>You can specify a set of annotators for a field using the <code>annotators</code> annotation.</p>
<p>Ex.:</p>
<pre><code class="language-yaml">  Gene:
    is_a: NamedThing
    id_prefixes:
      - HGNC
    annotations:
      annotators: gilda:, bioportal:hgnc-nr, obo:pr
</code></pre>
<p>The annotators are applied in order.</p>
<p>Additionally, when performing grounding, the following measures can be taken to improve accuracy:</p>
<ul>
<li>Specify the valid set of ID prefixes using <code>id_prefixes</code></li>
<li>Some vocabularies have structural IDs that are amenable to regexes, you can specify these using <code>pattern</code></li>
<li>You can make use of <code>values_from</code> slot to specify a <a href="https://linkml.io/linkml/schemas/enums.html#dynamic-enums">Dynamic Value Set</a></li>
<li>For example, you can constrain the set of valid locations for a gene product to be subclasses of <code>cellular_component</code> in GO or <code>cell</code> in CL</li>
</ul>
<p>Ex.:</p>
<pre><code class="language-yaml">classes:
  ...
  GeneLocation:
    is_a: NamedEntity
    id_prefixes:
      - GO
      - CL
    annotations:
      annotators: &quot;sqlite:obo:go, sqlite:obo:cl&quot;
    slot_usage:
      id:
        values_from:
          - GOCellComponentType
          - CellType

enums:
  GOCellComponentType:
    reachable_from:
      source_ontology: obo:go
      source_nodes:
        - GO:0005575 ## cellular_component
  CellType:
    reachable_from:
      source_ontology: obo:cl
      source_nodes:
        - CL:0000000 ## cell
</code></pre>
<h3 id="owl-exports">OWL Exports</h3>
<p>The <code>extract</code> command will let you export the results as OWL axioms, utilizing <a href="https://linkml.io/linkml-owl">linkml-owl</a> mappings in the schema.</p>
<p>Ex.:</p>
<pre><code class="language-bash">ontogpt extract -t recipe -i recipe-spaghetti.txt -o recipe-spaghetti.owl -O owl
</code></pre>
<p><a href="src/ontogpt/templates/recipe.yaml">src/ontogpt/templates/recipe.yaml</a> is an example schema that uses linkml-owl mappings.</p>
<p>See the <a href="Makefile">Makefile</a> for a full pipeline that involves using robot to extract a subset of FOODON
and merge in the extracted results. This uses <a href="https://github.com/hhursev/recipe-scrapers">recipe-scrapers</a>.</p>
<p>OWL output: <a href="tests/output/owl/merged/recipe-all-merged.owl">recipe-all-merged.owl</a></p>
<p>Classification:</p>
<p><img width="1329" alt="image" src="https://user-images.githubusercontent.com/50745/230427663-20d845e9-f1d5-490e-b1ad-cdccdd0dca70.png"></p>
<h2 id="web-application-setup">Web Application Setup</h2>
<p>There is a bare bones web application for running OntoGPT and viewing results.</p>
<p>Install the required dependencies by running the following command:</p>
<pre><code class="language-bash">poetry install -E web
</code></pre>
<p>Then run this command to start the web application:</p>
<pre><code class="language-bash">poetry run web-ontogpt
</code></pre>
<p>Note: The agent running uvicorn must have the API key set, so for obvious reasons don't host this publicly without authentication, unless you want your credits drained.</p>
<h2 id="ontogpt-limitations">OntoGPT Limitations</h2>
<ol>
<li>Non-deterministic</li>
<li>This relies on an existing LLM, and LLMs can be fickle in their responses</li>
<li>Coupled to OpenAI</li>
<li>You will need an OpenAI account to use their API. In theory any LLM can be used but in practice the parser is tuned for OpenAI's models</li>
</ol>
<h3 id="spindoctor-web-app">SPINDOCTOR web app</h3>
<p>To start:</p>
<pre><code>poetry run streamlit run src/ontogpt/streamlit/spindoctor.py
</code></pre>
<h3 id="huggingface-hub">HuggingFace Hub</h3>
<p>A select number of LLMs may be accessed through HuggingFace Hub. See the full list using <code>ontogpt list-models</code></p>
<p>Specify a model name with the <code>-m</code> option.</p>
<p>Example:</p>
<pre><code class="language-bash">ontogpt extract -t mendelian_disease.MendelianDisease -i tests/input/cases/mendelian-disease-sly.txt -m FLAN_T5_BASE
</code></pre>
<h3 id="using-local-models">Using local models</h3>
<p>OntoGPT supports using language models released by <a href="https://gpt4all.io/">GPT4All</a>.</p>
<p>Specify the name of a model when using the <code>extract</code> command with the <code>-m</code> or <code>--model</code> option and OntoGPT will retrieve the model.</p>
<p>For example:</p>
<pre><code>ontogpt --verbose extract -t mendelian_disease.MendelianDisease -i mendelian-disease-sly.txt -m ggml-gpt4all-j-v1.3-groovy
</code></pre>
<p>will download the <code>ggml-gpt4all-j-v1.3-groovy.bin</code> file, generate a prompt, and try that prompt against the specified model.</p>
<h2 id="citation">Citation</h2>
<p>SPIRES is described further in: Caufield JH, Hegde H, Emonet V, Harris NL, Joachimiak MP, Matentzoglu N, et al. Structured prompt interrogation and recursive extraction of semantics (SPIRES): A method for populating knowledge bases using zero-shot learning. </p>
<p>arXiv publication: http://arxiv.org/abs/2304.02711</p>
<h2 id="contributing">Contributing</h2>
<p>Contributions on recipes to test welcome from anyone! Just make a PR <a href="https://github.com/monarch-initiative/ontogpt/blob/main/tests/input/recipe-urls.csv">here</a>. See <a href="https://github.com/hhursev/recipe-scrapers">this list</a> for accepted URLs</p>
<h2 id="acknowledgements">Acknowledgements</h2>
<p>We gratefully acknowledge <a href="https://www.bosch.com/research">Bosch Research</a> for their support of this research project.</p>





                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": ".", "features": ["content.tabs.link"], "search": "assets/javascripts/workers/search.db81ec45.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="assets/javascripts/bundle.6df46069.min.js"></script>
      
    
  </body>
</html>